### ISSUE TITLE
UniqueValidator does not work with when one of targetAttribute has JSON type in DB.

### ISSUE THREAD

**Issue by floor12** [yiisoft/yii2#18289]
### What steps will reproduce the problem?

In case when one of target attributes has `JSON` type  in DB it will drop this validation to error: 'attr1 is invalid'.

```php
public function rules()
{
    return [
        ['attr1', 'unique', 'targetAttribute' => ['attr1',  'attr2', 'attr3', 'attr4', 'attr5']
    ];
}
`

### What is the expected result?

Unique validator make request to db to confirm conditions.

### What do you get instead?

Unique validator stop validating with error message  'attr1 is invalid' if any of attr(1,2,3,4 or 5) has `JSON` type in DB.

### Additional info

This problem happens because of this part in `UniqueValidator::validateAttribute()`:

```php
foreach ($rawConditions as $key => $value) {
    if (is_array($value)) {
        $this->addError($model, $attribute, Yii::t('yii', '{attribute} is invalid.'));
    return;
  }
  $conditions[] = [$key => $value];
}

````

This code will exit and add error to attribute if any of attribute has a array type. What for it was done? I want to make fix for this behavior, but firstly  i want to know the reason of this code, If somebody knows it. 
Now i have to extend `UniqueValidator` with overwriting `validateAttribute()` to work it in my current project, but i think i can make fix for this globaly. 

Now im going to deep history commits dive. Maybe i can found the reason there...

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.35
| PHP version      |  7.3
| Operating system | Linux

**Comment by floor12** [comment]
After some invesigation i found that this array check was made in 2012-2013 years. May be in this time AR attributes linked to db fields could not be arrays?

**Comment by bizley** [comment]
By default Yii is mapping JSON column's value to array using json_decode. I guess you want to check the JSON string here instead, right? This is quite unusual and I'm not sure if it requires adding changes in this validator.

**Comment by floor12** [comment]
Yes, i thinks its not bad idea to check schema after array check and if it JSON field -  just add json_encode to this attribute and check it like a string.  I can do this job with tests.
If nobody needs this fix except me, we can leave this problem. But it will be nice to add some information of this behaviour in core validators documentation....

**Comment by bizley** [comment]
To be honest I think your use case is very specific and so it should be done as a custom feature in your own code and not to be resolved in core. But I will leave it for discussion, at least for a while.

**Comment by samdark** [comment]
> After some invesigation i found that this array check was made in 2012-2013 years. May be at that time AR attributes linked to db fields could not be arrays?

You are correct. JSON support was added later.

Is there a good way to fix it?

**Comment by floor12** [comment]
I think something like this will fix this problem:
```php
$targetClass = $this->getTargetClass($model);
$db = $targetClass::getDb();
if (is_array($condition) && $db
   ->getSchema()
   ->getTableSchema($targetClass::tableName())
   ->getColumn($attributeName)
   ->type == Schema::TYPE_JSON) {
         $condition = json_encode($condition);
      }
```

But to prevent some code duplicates in UniqueValidator and make it more beauty and readble it needs some refactoring. It needs to move a couple of standaloen variables into the private attributes of the class and extract json check in a seperate method. And of course add some tests that I haven't looked into yet. 
I will came back with PR in a few days.

**Comment by bizley** [comment]
Wouldn't be easier just to remove the requirement for the value to not be array and then to check if arrays are the same? The check for JSON column would not be needed then.

**Comment by floor12** [comment]
I don't quite understand what you mean.

The main job of this validator is to create a single query (`::modelExists() method`) into the database trying to find a record that satisfies the condition. If a record is found, the validator generates an error.
What kind of comparison of 2 arrays are you talking about?

I see the only simpler solution to the problem in converting all the array to a live string, but it's hard for me to predict what consequences this might cause. Therefore, I decided to rely on the table diagram as the most reliable source of information.

Perhaps I am missing something.

**Comment by bizley** [comment]
I was referring to [this line](https://github.com/yiisoft/yii2/blob/master/framework/validators/UniqueValidator.php#L132) but yes, you are right, it's not enough anyway.

**Comment by floor12** [comment]
Despite the fact that work temporarily took all my time, I still want to return to this issue. The patch for the validator is ready, the problem is in testing it. The tests are not very simple + some mocks do not contain methods used in the validator. I want to complete this issue at the end.
